packages/plugin-hardhat/src/verify-proxy.ts

```diff
@@ -0,0 +1,527 @@
+import {
+  toCheckStatusRequest,
+  toVerifyRequest,
+} from '@nomiclabs/hardhat-etherscan/dist/src/etherscan/EtherscanVerifyContractRequest';
+import {
+  delay,
+  getVerificationStatus,
+  verifyContract,
+} from '@nomiclabs/hardhat-etherscan/dist/src/etherscan/EtherscanService';
+
+import {
+  getTransactionByHash,
+  getImplementationAddress,
+  getBeaconAddress,
+  getImplementationAddressFromBeacon,
+  UpgradesError,
+  getAdminAddress,
+  isTransparentOrUUPSProxy,
+  isBeaconProxy,
+  isEmptySlot,
+} from '@openzeppelin/upgrades-core';
+import artifactsBuildInfo from '@openzeppelin/upgrades-core/artifacts/build-info.json';
+
+import { HardhatRuntimeEnvironment, RunSuperFunction } from 'hardhat/types';
+
+import ERC1967Proxy from '@openzeppelin/upgrades-core/artifacts/@openzeppelin/contracts/proxy/ERC1967/ERC1967Proxy.sol/ERC1967Proxy.json';
+import BeaconProxy from '@openzeppelin/upgrades-core/artifacts/@openzeppelin/contracts/proxy/beacon/BeaconProxy.sol/BeaconProxy.json';
+import UpgradeableBeacon from '@openzeppelin/upgrades-core/artifacts/@openzeppelin/contracts/proxy/beacon/UpgradeableBeacon.sol/UpgradeableBeacon.json';
+import TransparentUpgradeableProxy from '@openzeppelin/upgrades-core/artifacts/@openzeppelin/contracts/proxy/transparent/TransparentUpgradeableProxy.sol/TransparentUpgradeableProxy.json';
+import ProxyAdmin from '@openzeppelin/upgrades-core/artifacts/@openzeppelin/contracts/proxy/transparent/ProxyAdmin.sol/ProxyAdmin.json';
+
+import { keccak256 } from 'ethereumjs-util';
+
+import debug from './utils/debug';
+import { callEtherscanApi, EtherscanAPIConfig, getEtherscanAPIConfig, RESPONSE_OK } from './utils/etherscan-api';
+
+/**
+ * Hardhat artifact for a precompiled contract
+ */
+interface ContractArtifact {
+  contractName: string;
+  sourceName: string;
+  abi: any;
+  bytecode: any;
+}
+
+/**
+ * A mapping from a contract artifact to the event that it logs during construction.
+ */
+interface ContractEventMapping {
+  artifact: ContractArtifact;
+  event: string;
+}
+
+/**
+ * Types of proxy-related contracts and their corresponding events.
+ */
+interface ContractEventsMap {
```

Comment:
I find the type names confusing. I would rename:

- `ContractEventsMap` -> `VerifiableContracts`
- `ContractEventsMapping` -> `VerifiableContractInfo`